suite-common/suite-utils/src/device.ts

```diff
@@ -360,6 +360,32 @@ export const getDeviceInstances = (
         }) as AcquiredDevice[];
 };
 
+/**
+ * Returns all device instances sorted by `instance` field for all connected and remembered devices
+ * * @param {TrezorDevice[]} devices
+ * @returns {AcquiredDevice[][]}
+ */
+export const getDeviceInstancesGroupedByDeviceId = (
+    devices: TrezorDevice[],
+): AcquiredDevice[][] => {
+    const deviceIds: string[] = [];
+    const deviceGroups: AcquiredDevice[][] = [];
+
+    for (const device of devices) {
+        if (!device.features || !device.id || deviceIds.includes(device.id)) continue;
+
+        if (deviceIds.includes(device.id)) {
+            continue;
+        }
+
+        deviceIds.push(device.id);
+
+        deviceGroups.push(getDeviceInstances(device, devices));
+    }
+
+    return deviceGroups;
+};
```

Comment:
I know it's a matter of taste, but I almost always found declarative way easier to read and think about than mutating things. But of course it's up to you here.
```suggestion
): AcquiredDevice[][] => devices.reduce((deviceGroups, device) => {
    if (!device.features || !device.id) {
        return deviceGroups;
    }

    const existingGroupIndex = deviceGroups.findIndex(group => group[0].id === device.id);
    if (existingGroupIndex === -1) {
        // If the device ID is not yet in the accumulator, add a new group
        const newGroup = getDeviceInstances(device, devices);
        if (newGroup.length > 0) {
            deviceGroups.push(newGroup);
        }
    }

    return deviceGroups;
}, [] as AcquiredDevice[][]);
```

Reply:
b38e5951ad